pkg/protocol/engine/attestation/slotattestation/manager.go

```diff
@@ -0,0 +1,249 @@
+
+		e.Storage.Settings().HookInitialized(func() {
+			m.commitmentMutex.Lock()
+			m.lastCommittedSlot = e.Storage.Settings().LatestCommitment().ID().Index()
+			m.lastCumulativeWeight = e.Storage.Settings().LatestCommitment().Commitment().CumulativeWeight
+			m.commitmentMutex.Unlock()
+		})
+
+		return m
+	})
+}
+
+func NewManager(attestationCommitmentOffset iotago.SlotIndex, bucketedStorage func(index iotago.SlotIndex) kvstore.KVStore, committeeFunc func() *account.Accounts[iotago.AccountID, *iotago.AccountID]) *Manager {
+	m := &Manager{
+		attestationCommitmentOffset: attestationCommitmentOffset,
+		committeeFunc:               committeeFunc,
+		bucketedStorage:             bucketedStorage,
+		futureAttestations:          memstorage.NewIndexedStorage[iotago.SlotIndex, iotago.AccountID, *iotago.Attestation](),
+		pendingAttestations:         memstorage.NewIndexedStorage[iotago.SlotIndex, iotago.AccountID, *iotago.Attestation](),
+	}
+	m.TriggerConstructed()
+
+	return m
+}
+
+func (m *Manager) Shutdown() {
+	if err := m.writeToDisk(); err != nil {
+		panic(err)
+	}
+	m.TriggerStopped()
+}
+
+// Get returns the attestations that are included in the commitment of the given slot.
+// If attestationOffset=3 and commitment is 10, then the returned attestations are blocks from 7 to 10 that commit to at least 7.
+func (m *Manager) Get(index iotago.SlotIndex) (*ads.Map[iotago.AccountID, iotago.Attestation, *iotago.AccountID, *iotago.Attestation], error) {
+	m.commitmentMutex.RLock()
+	defer m.commitmentMutex.RUnlock()
+
+	if index > m.lastCommittedSlot {
+		return nil, errors.Errorf("slot %d is newer than last committed slot %d", index, m.lastCommittedSlot)
+	}
+	cutoffIndex, isValid := m.computeAttestationCommitmentOffset(index)
+	if !isValid {
+		return nil, errors.Errorf("slot %d is smaller than attestation cutoffIndex %d thus we don't have attestations", index, cutoffIndex)
+	}
+
+	return m.adsMapStorage(cutoffIndex)
+}
+
+// AddAttestationFromBlock adds an attestation from a block to the future attestations (beyond the attestation window).
+func (m *Manager) AddAttestationFromBlock(block *blocks.Block) {
+	// Only track attestations of active committee members.
+	if _, exists := m.committeeFunc().Get(block.Block().IssuerID); !exists {
+		return
+	}
+
+	m.commitmentMutex.RLock()
+	defer m.commitmentMutex.RUnlock()
+
+	// We only care about attestations that are newer than the last committed slot.
+	if block.ID().Index() <= m.lastCommittedSlot {
+		return
+	}
+
+	newAttestation := iotago.NewAttestation(block.Block())
+
+	// We keep only the latest attestation for each committee member.
+	m.futureAttestations.Get(block.ID().Index(), true).Compute(block.Block().IssuerID, func(currentValue *iotago.Attestation, exists bool) *iotago.Attestation {
+		if !exists {
+			return newAttestation
+		}
+
+		// Replace the attestation only if the new one is greater.
+		if newAttestation.Compare(currentValue) == 1 {
```

Comment:
You are only comparing the issuing time of the block here, not what they commit to. So in theory a newer block that commits to something older would be considered newer. We already had a chat about this, but I don't remember why we decided to remove the commitment from the comparison.